### PR TITLE
Parsing JSON Reports to allow Ignored Warnings to be displayed

### LINKED ISSUE #8: Displaying Ignored Warnings
Hey Everyone,

My team has been using this plugin on our Jenkins server for a while, and we think that it would be very useful to have a display for the current ignored brakeman warnings. Very similar to the way the plugin already displays the vulnerabilities. This would be very helpful in seeing all the different warnings that are being ignored, in an easily readable way, since the list can get long in cases.

I think another useful functionality to go along with this could also be a threshold limiting these ignored warnings (ex. > x number of ignored warnings makes the build unstable). I think this could definitely provoke people to fix these security vulnerabilities instead of just marking them as "ignored" when they find them. That being said, a display for the ignored warnings would definitely be the most useful.

I've delved into some of the work that goes into using analysis-core to parse these files (brakeman-output & brakeman.ignore) and add the different warnings that are found. I also found it fairly easy to parse the brakeman.ignore files, and add annotations for them. The problem I found with that, is that the reports are generated from a collection of all these annotations, and do not differentiate from the different objects that you create annotations with. Ex:

`project.addAnnotation(Warning.new(args));`
`project.addAnnotation(IgnoredWarning.new(args));`

I was wondering if there was another way to accomplish this differentiation of information? And/or if this is a feature that you all would be potentially interested in adding?

I am definitely willing to create a PR for this, if you all are open to outside contributions!

Any help would be greatly appreciated! Thanks!

### PR DESCRIPTION
**Pull request by tmobaird**
Sorry for this taking so long, but this ended up being a fairly large PR. This PR is in reference to #8 about reading JSON reports and displaying ignored warnings (while remaining backwards compatible to read the tabs reports as well). My strategy to accomplish this was to check the report file content and check if it matches a JSONObject. If it does, it will scan the file with the JSON strategy, if not it will scan the file based on the tabs strategy. 

I also had to make sure that these new ignored warnings that can be added would not affect the status of a build (ex. a build won't fail if a threshold is set for 0 warnings, and a brakeman report includes 1 ignored warning). This would've probably made a lot of developers mad if their builds started failing because of this; so I made `evaluateStatus` only factor for non-ignored annotations/warnings.

Everything is working fine for me in two different local Jenkins instances!

Here are some more details about what this PR includes:
- Ability to use JSON reports (as well as tabs)
- Can read ignored warnings in JSON reports
- Annotations with category "Ignored" don't factor into threshold evaluations
- Ignored Warnings Total on build status page
- Ignored Warnings added to Reports screens and Trend Graphs
- Error handling for invalid JSON reports

This closes #8.

Feel free to let me know if there are any suggestions or concerns!

### CODE REVIEW

**Comment by ConstantineXVI** [comment]
I'll review this [hopefully] sometime this weekend.

**Comment by tmobaird** [comment]
@ConstantineXVI sounds good, thanks!

File: src/main/java/hudson/plugins/brakeman/scanners/BrakemanJSONScanner.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+                append(row.getString("message")).
+                append(": ").
+                append(row.getString("code"));
+            Priority priority = checkPriority(row.getString("confidence"));
+
+            project.addAnnotation(new Warning(fileName, getStart(line), getEnd(line), type, category, message.toString(), priority));
+        }
+    }
+
+    private String getCategory(String filterType) {
```
Comment (ConstantineXVI) [inline src/main/java/hudson/plugins/brakeman/scanners/BrakemanJSONScanner.java]: Likewise on the enums here.

File: pom.xml [hunk 2]
```diff
@@ -55,6 +55,11 @@
       <artifactId>workflow-step-api</artifactId>
       <version>1.15</version>
     </dependency>
+    <dependency>
+      <groupId>org.json</groupId>
+      <artifactId>json</artifactId>
+      <version>20090211</version>
```
Comment (tmobaird) [inline pom.xml]: That's my bad. I meant to update this before I created the PR. I believe it was just the most popular version, which is why I started with this. I'll update this to a newer version

File: src/main/java/hudson/plugins/brakeman/BrakemanPublisher.java [hunk 3]
```diff
@@ -91,67 +91,71 @@ public Action getProjectAction(final AbstractProject<?, ?> project) {
...
+        return scanner;
+    }
+
+	/**
+	 * Determines the file type based upon the contents.
+	 *
+     * @param content
+	 * @return result
+	 */
+	protected String getFileType(String content) {
```
Comment (tmobaird) [inline src/main/java/hudson/plugins/brakeman/BrakemanPublisher.java]: That sounds good to me. I'll update asap 👍

File: src/main/java/hudson/plugins/brakeman/scanners/BrakemanJSONScanner.java (hunk 1 above)
Comment (tmobaird) [inline src/main/java/hudson/plugins/brakeman/scanners/BrakemanJSONScanner.java]: 👍

**Comment by tmobaird** [comment]
Ok all issues with this PR **should** be resolved now. Here's a summary of the changes:
- Removed org.json dependency from pom, not needed because net.sf.json is available to you
- Changed org.json imports to net.sf.json, and refactored a little bit of the JSON parsing code to use this library's [net.sf.json] code
- Switched a couple methods to return an enum, per suggestion

Let me know if there's anything else that is needed.

**Comment by tmobaird** [comment]
@ConstantineXVI any progress on this?

**Comment by ConstantineXVI** [comment]
Sorry, I've been swamped lately. I'm hoping to finish this up over the holiday.

**Comment by tmobaird** [comment]
@ConstantineXVI no worries, I was just curious. Thanks!

**Comment by tmobaird** [comment]
@ConstantineXVI thanks! 👍

**Pull request merged**